src/pcre2_jit_compile.c

```diff
@@ -7685,8 +7685,233 @@ return cc;
     } \
   charoffset = (value);
 
+#define READ_FROM_CHAR_LIST(destination) \
+  if (list_ind <= 1) \
+    { \
+    destination = *(uint16_t*)next_char; \
+    next_char += 2; \
+    } \
+  else \
+    { \
+    destination = *(uint32_t*)next_char; \
+    next_char += 4; \
+    }
+
+#define XCLASS_LOCAL_RANGES_SIZE 32
+#define XCLASS_LOCAL_RANGES_LOG2_SIZE 5
+
 static PCRE2_SPTR compile_char1_matchingpath(compiler_common *common, PCRE2_UCHAR type, PCRE2_SPTR cc, jump_list **backtracks, BOOL check_str_ptr);
 
+typedef struct xclass_stack_item {
+  sljit_u32 first_item;
+  sljit_u32 last_item;
+  struct sljit_jump *jump;
+} xclass_stack_item;
+
+typedef struct xclass_ranges {
+  size_t range_count;
+  /* Pointer to ranges. A stack area is provided when a small buffer is enough. */
+  uint32_t *ranges;
+  uint32_t local_ranges[XCLASS_LOCAL_RANGES_SIZE * 2];
+  /* Stack size must be log2(ranges / 2). */
+  xclass_stack_item *stack;
+  xclass_stack_item local_stack[XCLASS_LOCAL_RANGES_LOG2_SIZE];
+} xclass_ranges;
+
+static void xclass_compute_ranges(compiler_common *common, PCRE2_SPTR cc, xclass_ranges *ranges)
+{
+DEFINE_COMPILER;
+size_t range_count = 0, est_range_count;
+size_t total_size, est_stack_size, tmp;
```

Comment:
It is used once, so I removed it. Why shadowing is a problem btw? The whole point of local scoping is shadowing.

Reply:
Philip doesn't like it, and I think it is because with functions being so large, it is more difficult to keep track of their values when they shadow.

Of course nothing prevents you using the same variable in smaller scopes, like I was suggesting for "average", so I think by being strict about it, also pushes for smaller scopes overall that can only help long term with maintenance.

Reply:
Yes, indeed, I don't like shadow variables because they can be confusing if you are reading unfamiliar code. Suppose you see something like this:

```
       X = some value;
       more code...
       }
    code involving X 
```
You would probably expect the second X to be the same variable as the first X and then when it doesn't make sense you have to look back up the code to figure it out. I do know that I tend to write very large functions, which makes this worse. I blame it on having started out when compilers were not clever and calling a function was more costly than it is now. (And computers were slower, of course. :-)